tests/function_modifiers/test_macros.py

```diff
@@ -680,11 +679,81 @@ def test_pipe_output_end_to_end_1():
     assert result["chain_2_using_pipe_output"] == result["chain_2_not_using_pipe_output"]
 
 
-def test_pipe_output_end_to_end_2():
+def mutator_function(input_1: int, input_2: int) -> int:
+    return input_1 + input_2
+
+
+def test_mutate_convert_callable_to_applicable():
+    decorator = mutate(result_from_downstream_function)
+
+    assert len(decorator.remote_applicables) == 1
+    remote_applicable = decorator.remote_applicables[0]
+    assert isinstance(remote_applicable, Applicable)
+    assert remote_applicable.fn is None
+    assert remote_applicable.target_fn == result_from_downstream_function
+
+
+def test_mutate_applies_underscore():
+    correct_name = "".join(["_", mutator_function.__name__])
+    decorated_function = mutate(result_from_downstream_function, input_2=0)(mutator_function)
+
+    assert correct_name == decorated_function.__name__
+
+
+def test_mutate_restricted_to_same_module():
+    decorator = mutate(tests.resources.mutate.f_of_interest)
+
+    with pytest.raises(hamilton.function_modifiers.macros.NotSameModuleError):
+        decorator.validate_same_module(mutator_function)
+
+
+def test_mutate_global_kwargs():
+    decorator = mutate(apply_to(result_from_downstream_function), input_2=17)
+    remote_applicable = decorator.remote_applicables[0]
+
+    pipe_step = decorator._create_step(
+        mutating_fn=mutator_function, remote_applicable_builder=remote_applicable
+    )
+    assert pipe_step.kwargs["input_2"] == 17
+
+
+def test_mutate_local_kwargs_override_global_ones():
+    decorator = mutate(apply_to(result_from_downstream_function, input_2=13), input_2=17)
+    remote_applicable = decorator.remote_applicables[0]
+
+    pipe_step = decorator._create_step(
+        mutating_fn=mutator_function, remote_applicable_builder=remote_applicable
+    )
+    assert pipe_step.kwargs["input_2"] == 13
+
+
+def test_mutate_end_to_end_simple():
+    dr = driver.Builder().with_config({"calc_c": True}).build()
+
+    dr = (
+        driver.Builder()
+        .with_modules(tests.resources.mutate)
```

Comment:
So there's a bit of an odd issue here. Specifically, `mutate` will mark the modules (and leave a mark). Thus calling it a second time (for instance through pmultiple tests) might mess it up slightly... I'm imagining you haven't hit that, but you miay want to mess with fixtures/whatnot, or do a dynamic re-import just to be sure.

Also worth documenting.

Reply:
It actually would show up in the ones above, not this